src/runtime_src/core/tools/xbutil2/OO_P2P.cpp

```diff
@@ -17,12 +17,244 @@
+           << pattern << "' got '" << (*itr) << "'";
+      throw xrt_core::system_error(EIO, ostr.str());
+    }
+  }
+}
+
+static void
+chunk(xrt_core::device* device, char* boptr, uint64_t dev_addr, uint64_t size)
+{
+  char patternA = 'A';
+  char patternB = 'B';
+
+  auto mem = xrt_core::aligned_alloc(xrt_core::getpagesize(), size);
+  auto buf = static_cast<char*>(mem.get());
+
+  fill_with_stride(buf, buf+size, patternA);
+  try {
+    device->unmgd_pwrite(buf, size, dev_addr);
+  }
+  catch (const std::exception&) {
+    std::ostringstream ostr;
+    ostr << "Error writing 0x" << std::hex << size
+         << " bytes to 0x" << dev_addr << std::dec;
+    throw xrt_core::system_error(EIO, ostr.str());
+  }
+  cmp_with_stride(boptr, boptr+size, patternA);
+
+  fill_with_stride(boptr, boptr+size, patternB);
+  try {
+    device->unmgd_pread(buf, size, dev_addr);
+  }
+  catch (const std::exception&) {
+    std::ostringstream ostr;
+    ostr << "Error reading 0x" << std::hex << size
+         << " bytes from 0x" << dev_addr << std::dec;
+    throw xrt_core::system_error(EIO, ostr.str());
+  }
+  cmp_with_stride(buf, buf+size, patternB);
+}
+
+static void
+bank(xrt_core::device* device, int memidx, uint64_t addr, uint64_t size)
+{
+  const size_t chunk_size = 16 * 1024 * 1024;
+
+  auto bo = xrt::bo(device->get_device_handle(), size, XCL_BO_FLAGS_P2P, memidx);
+  auto boptr = bo.map<char*>();
+
+  for (size_t c = 0, ci = 0; c < size; c += chunk_size, ++ci) {
+    try {
+      chunk(device, boptr + c, addr + c, chunk_size);
+    }
+    catch (const std::exception& ex) {
+      std::ostringstream ostr;
+      ostr << ex.what() << "\n"
+           << "Error p2p testing at offset 0x" << std::hex << c
+           << " on memory index " << std::dec << memidx;
+      throw xrt_core::system_error(EINVAL, ostr.str());
+    }
+
+    if (ci % (size / chunk_size / 16) == 0)
+      std::cout << "." << std::flush;
+  }
+}
+
+// p2ptest
+static void
+test(xrt_core::device* device)
+{
+  // lock xclbin
+  auto uuid = xrt::uuid(xrt_core::device_query<xrt_core::query::xclbin_uuid>(device));
+  auto at_exit = [] (auto device, auto uuid) { device->close_context(uuid.get(), -1); };
+  xrt_core::scope_guard<std::function<void()>> g(std::bind(at_exit, device, uuid));
+  device->open_context(uuid.get(), -1, true);
+
+  // p2p must be enabled
+  p2p_enabled_or_error(device);
+
+  // Get first memory bank supported by P2P
+  auto mt_raw = xrt_core::device_query<xrt_core::query::mem_topology_raw>(device);
+  auto mt = reinterpret_cast<const mem_topology*>(mt_raw.data());
+  if (!mt)
+    throw xrt_core::system_error(EINVAL, "mem_topology is invalid, cannot validate p2p");
+
+  // support memory types
+  // p2p is not supported for DDR on u280
+  auto vbnv = xrt_core::device_query<xrt_core::query::rom_vbnv>(device);
+  std::vector<std::string> supported = { "HBM", "bank" };
+  if (vbnv.find("_U280_") == std::string::npos)
```

Comment:
So sad....there needs to be a better way, but unfortunately that doesn't look to be the case.